tasks/add-unique-granuleID/tests/index.js

```diff
@@ -0,0 +1,80 @@
+const test = require('ava');
+const { assignUniqueIds } = require('../dist/src');
+
+test('assignUniqueIds assigns unique granule IDs and preserves producerGranuleId', async (t) => {
+  const hashDepth = 6;
+
+  const event = {
+    input: {
+      granules: [
+        { granuleId: 'granule1', collectionId: 'collection1' },
+        { granuleId: 'granule2', collectionId: 'collection2' },
+      ],
+    },
+    config: {
+      hashDepth,
+    },
+  };
+
+  const originalGranules = structuredClone(event.input.granules);
+  const result = await assignUniqueIds(event, {});
+
+  t.is(result.granules.length, 2, 'Output should have the same number of granules');
+  result.granules.forEach((granule, index) => {
+    t.is(granule.producerGranuleId, originalGranules[index].granuleId, 'Should preserve producerGranuleId as original granuleID value');
+    t.true(granule.granuleId.startsWith(`${granule.producerGranuleId}_`), 'Should assign a unique granuleId');
+    t.is(granule.granuleId.split('_')[1].length, hashDepth, 'Hash length should match the configured hashDepth');
+  });
+});
+
+test('assignUniqueIds ignores granules that already have producerID assigned', async (t) => {
+  const event = {
+    input: {
+      granules: [
+        { granuleId: 'granule1', collectionId: 'collection1', producerGranuleId: 'existingGranuleId1' },
+        { granuleId: 'granule2', collectionId: 'collection2', producerGranuleId: 'existingGranuleId2' },
+      ],
+    },
+  };
+
+  const result = await assignUniqueIds(event, {});
+  const originalGranules = structuredClone(event.input.granules);
+
+  t.is(result.granules.length, 2, 'Output should have the same number of granules');
+  result.granules.forEach((granule, index) => {
+    t.is(granule.producerGranuleId, originalGranules[index].producerGranuleId, 'Should preserve existing producerGranuleId');
+    t.is(granule.granuleId, originalGranules[index].granuleId, 'Should not change granuleId if producerGranuleId is already set');
```

Comment:
Ah that's good we're explicitly asserting this because I'm not sure what we want here 🤔

Reply:
_is_ there additional thought that needs to be given to this logic in a later ticket? I assume so because whatever we decide for our migration that populates this field will impact this. e.g. if DAACs have producerGranuleId populated in a migration but do NOT have unique granuleIds, they'd still need to unique-ify.

In any case the assertion here is correct for now. Just wanting to make sure that I'm on the same page that this is open to change later.

Reply:
`is there additional thought that needs to be given to this logic in a later ticket? `

I'm ... not aware of that, other than if updates need to be made to conform to an integration test assertion this isn't meeting.   

`if DAACs have producerGranuleId populated in a migration but do NOT have unique granuleIds, they'd still need to unique-ify.`

I'm not sure  that's true - existing granules shouldn't need to make a new unique ID, we're not requiring them to do this at all, granules already in the system *have* a unique ID, the problem is newer granules that might collide. 

The weird edge case here *should* be only if someone attempts to re-run a granule that doesn't get that value populated (we choose not to migrate) and it *inadvertently* uniqueifies a granule.

Reply:
Yeah, this is answered by our previous conversation. There isn't a use case we're aware of that makes this a problem.